python-package/lightgbm/basic.py

```diff
@@ -560,8 +563,14 @@ def predict(self, data, num_iteration=-1,
         if pred_contrib:
             predict_type = C_API_PREDICT_CONTRIB
         int_data_has_header = 1 if data_has_header else 0
-        if num_iteration > self.num_total_iteration:
-            num_iteration = self.num_total_iteration
+        if start_iteration < 0:
+            start_iteration = 0
+        if start_iteration >= self.num_total_iteration:
+            start_iteration = self.num_total_iteration - 1
+        if num_iteration <= 0:
+            num_iteration = self.num_total_iteration - start_iteration
+        if num_iteration + start_iteration > self.num_total_iteration:
+            num_iteration = self.num_total_iteration - start_iteration
```

Comment:
Can we perform some of these checks at cpp side, like it is done for `num_iteration`?

https://github.com/microsoft/LightGBM/blob/a9f5654b1d5d10ec711f83807c4e94e7403918f5/src/boosting/gbdt.h#L216-L217